source/manual/test-and-build-a-project-with-github-actions.html.md

```diff
@@ -27,22 +27,64 @@ You should name the file used for configuring the CI workflow `ci.yml`.
 
 ### When the CI workflow should run
 
-The workflow should be configured to run on [`push`][push-event] and
-[`pull_request`][pull-request-event] events. The `push` event occurs for every
-push to the repository and this will generate a build against the branch
-that was pushed. The `pull_request` event occurs whenever a pull request
-is opened and it is triggered again any time the pull request changes (such
-as extra commits or rebasing), these differ to the `push` event by merging in
-the main branch before running operations. Pull requests from forks of a
-repo will only trigger the `pull_request` event.
+The workflow should be configured to run when there is a [`push`][push-event]
+to the default branch (typically "main") and when a
+[`pull_request`][pull-request-event] opened. This means CI runs when a branch
+is merged (the `push` to main) and it runs against any changes introduced
+in a pull request. We prefer to run against a `pull_request` event to all
+`push` events, as this allows external contributors to have pull requests
+tested and because `pull_request` merges the changes with
+the repository default branch.
```

Comment:
I don't understand this last bit. Don't GitHub actions run against the branch as-is, whether or not the trigger was `push` or `pull_request`?

Reply:
So I only found out about the details of this recently myself (thanks to @kevindew for putting me onto it) — pull_request uses the `refs/pull/<pr_number>/merge` reference that GitHub uses for setting up the merge. (I believe this is essentially the PR branch rebased onto the destination branch, but don't quote me on that.) Compare https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request with https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push

If the PR branch is up-to-date with main then there should be no practical difference, but if they've diverged significantly and "Require branches to be up-to-date before merging" isn't enabled then it could mean you're testing against an old version of the rest of the codebase.

Reply:
Thanks for the explanation! Perhaps we should amend that sentence to link back to your comment? 🙏